### PR TITLE
Refactor get Name or Id

### PR DESCRIPTION
**Pull request by anistouri**
Signed-off-by: TOURI ANIS <anis-1.touri@rte-france.com>

### CODE REVIEW

File: src/components/voltage-level-choice.js [hunk 1]
```diff
@@ -10,16 +10,13 @@ import makeStyles from '@mui/styles/makeStyles';
...
 import ListItemText from '@mui/material/ListItemText';
 import Button from '@mui/material/Button';
 import Typography from '@mui/material/Typography';
-import { useSelector } from 'react-redux';
 import { getNominalVoltageColor } from '../utils/colors';
-import { PARAM_USE_NAME } from '../utils/config-params';
-import { getNameOrId } from './diagrams/singleLineDiagram/utils';
+import { useNameOrId } from './util/equipmentInfosHandler';
 
-const useStyles = makeStyles((theme) => ({
```
Comment (sBouzols) [inline src/components/voltage-level-choice.js:22]: do not remove theme here, it's better to use theme.spacing(1) etc... than raw values for example, even if it's not use for the moment...
I understand it's a cleaner code if not used, maybe in another PR ?

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 2]
```diff
@@ -100,8 +102,7 @@ const useDisplayView = (network, studyUuid, currentNode) => {
             function createSubstationSLD(substationId, state) {
                 const substation = network.getSubstation(substationId);
                 if (!substation) return;
-                let name =
-                    useName && substation.name ? substation.name : substationId;
+                let name = getNameOrId(substation?.name, substationId);
```
Comment (sBouzols) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:105]: maybe rename the variable ?
`let label = ...` because it could be id or name here...

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 3]
```diff
@@ -121,7 +122,7 @@ const useDisplayView = (network, studyUuid, currentNode) => {
             function createVoltageLevelSLD(vlId, state) {
                 const vl = network.getVoltageLevel(vlId);
                 if (!vl) return;
-                let name = useName ? vl.name : vlId;
+                let name = getNameOrId(vl?.name, vlId);
```
Comment (sBouzols) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:125]: same `let label`

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 4]
```diff
@@ -224,7 +225,7 @@ export function SingleLineDiagramPane({
     const classes = useStyles();
 
     const handleUpdateSwitchState = useCallback(
-        (breakerId, open, switchElement) => {
```
Comment (sBouzols) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:227]: keep it even if not used, it's pass throught SingleLineDiagram:onBreakClick.
Or clean this in another PR.

File: src/components/graph/menus/modification-list-item.js [hunk 5]
```diff
@@ -93,7 +91,7 @@ export const ModificationListItem = ({
             function getVoltageLevelLabel(vlID) {
                 if (!vlID) return '';
                 const vl = network.getVoltageLevel(vlID);
-                if (vl) return vl[useName ? 'name' : 'id'] || vlID;
+                if (vl) return getNameOrId(vl['name'], vl['id']) ?? vlID;
```
Comment (sBouzols) [inline src/components/graph/menus/modification-list-item.js:94]: remove last part ` ?? vlID` it's done by your function, and use the same syntax than elsewhere :
`getNameOrId(vl.name, vl.id)`

File: src/components/menus/base-equipment-menu.js [hunk 6]
```diff
@@ -18,13 +18,9 @@ import { NestedMenuItem } from 'mui-nested-menu';
...
 import { useSelector } from 'react-redux';
-import { PARAM_USE_NAME } from '../../utils/config-params';
-import {
-    getNameOrId,
-    getSubstationNameOrId,
-} from '../diagrams/singleLineDiagram/utils';
+import { useNameOrId } from '../util/equipmentInfosHandler';
 
-const useStyles = makeStyles((theme) => ({
+const useStyles = makeStyles(() => ({
```
Comment (sBouzols) [inline src/components/menus/base-equipment-menu.js:23]: keep theme or clean it in another PR

File: src/components/util/equipmentInfosHandler.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+import { useCallback } from 'react';
+import { useSelector } from 'react-redux';
+import { PARAM_USE_NAME } from '../../utils/config-params';
+
+export const useNameOrId = () => {
+    const useName = useSelector((state) => state[PARAM_USE_NAME]);
+    const getNameOrId = useCallback(
+        (name, id) => {
```
Comment (sBouzols) [inline src/components/util/equipmentInfosHandler.js:15]: Why not taking only an `equipmentObject` as parameter and use `equipmentObject?.id`, `equipmentObject?.name` inside this function ?

File: package.json [hunk 8]
```diff
@@ -15,6 +15,7 @@
         "@reduxjs/toolkit": "^1.2.3",
         "@svgdotjs/svg.js": "^3.0.12",
         "@svgdotjs/svg.panzoom.js": "^2.1.2",
+        "@testing-library/react": "13.4.0",
```
Comment (sBouzols) [inline package.json:18]: Not sure... we already have jest throught CRA, why adding another testing framework ?

Comment (sBouzols) [reply package.json:18]: And there is no pakage-lock.json diff pushed then it won't work anyway.

**Review by sBouzols — CHANGES_REQUESTED** [review]
Some comments and
Some other parts to clean :
* PARAM_USE_NAME used in src/components/menus/line-menu.js
* PARAM_USE_NAME used in src/components/util/search-matching-equipments.js, you must pass the new function returned by your custom hook as parameter of `makeItems` to use it inside.
* PARAM_USE_NAME used in study-pane : `useName` is get only to be pass as prop to `NetworkMapTab`, then only to be pass as props to `NetworkMap`, maybe we could do something to clean this. Get it only in `NetworkMap` and do what needed

File: src/components/util/equipmentInfosHandler.js (hunk 7 above)
Comment (anistouri) [reply src/components/util/equipmentInfosHandler.js:15]: because in some cases we don't have the same parameter name. 
For example : `equipment?.substationName,  equipment?.substationId`

**Review by EstherDarkish — APPROVED** [review]

File: src/components/util/search-matching-equipments.js [hunk 9]
```diff
@@ -37,15 +36,14 @@ export const useSearchMatchingEquipments = (
...
+                        getParameter,
                         inUpstreamBuiltParentNode,
                         equipmentType
                     )
                         .then((infos) => {
                             if (searchTerm === lastSearchTermRef.current) {
-                                setEquipmentsFound(
-                                    makeItems(infos, useNameLocal)
-                                );
+                                const label = getNameOrId();
```
Comment (sBouzols) [inline src/components/util/search-matching-equipments.js:45]: you means `getParameter()` here I think no ?

File: src/components/util/equipmentInfosHandler.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+                return useName && name != null && name.trim() !== ''
+                    ? name
+                    : infos?.id;
+            }
+            return null;
+        },
+        [useName]
+    );
+
+    const getParameter = useName ? 'name' : 'id';
```
Comment (sBouzols) [inline src/components/util/equipmentInfosHandler.js:27]: getParameter function name is a bit too generic here I think.
Maybe something like `getUseNameParameterKey`

File: src/utils/rest-api.js [hunk 11]
```diff
@@ -577,7 +577,7 @@ export function fetchEquipmentsInfos(
     studyUuid,
     nodeUuid,
     searchTerm,
-    usesName,
+    getParameter,
```
Comment (sBouzols) [inline src/utils/rest-api.js:580]: it's a string here not a function, then name it `useNameParameterKey` ?

File: src/components/util/search-matching-equipments.js (hunk 9 above)
Comment (sBouzols) [reply src/components/util/search-matching-equipments.js:45]: `setEquipmentsFound(makeItems(infos, getUseNameParameterKey()));`

File: src/components/network/network-map.js [hunk 12]
```diff
@@ -346,7 +343,7 @@ const NetworkMap = (props) => {
                 network: props.network,
                 updatedLines: props.updatedLines,
                 geoData: props.geoData,
-                useName: props.useName,
+                useName: useName,
```
Comment (sBouzols) [inline src/components/network/network-map.js:346]: It seems we don't need useName in LineLayer... maybe we could clean this as well ?

File: src/components/network/network-map.js [hunk 13]
```diff
@@ -336,6 +332,7 @@ const NetworkMap = (props) => {
                 onHover: ({ object, x, y }) => {
                     setCursorType(object ? 'pointer' : 'grab');
                 },
+                getNameOrId: getNameOrId,
```
Comment (sBouzols) [inline src/components/network/network-map.js:335]: If we pass `getNameOrId` in `SubstationLayer` which changes each time the parameter `useName` changes then maybe we could avoid tu pass `useName` as props in `SubstationLayer` and trigger label refresh with `getNameOrId`

File: src/components/menus/line-menu.js [hunk 14]
```diff
@@ -85,11 +81,12 @@ const withLineMenu =
 
         const getLineDescriptor = useCallback(
             (voltageLevelId) => {
-                return displayUseName
-                    ? getNameOrId(network.getVoltageLevel(voltageLevelId))
-                    : voltageLevelId;
+                return getNameOrId({
```
Comment (sBouzols) [inline src/components/menus/line-menu.js:84]:
```
const vl = network.getVoltageLevel(voltageLevelId);
return getNameOrId(vl);
```

File: src/components/menus/base-equipment-menu.js [hunk 15]
```diff
@@ -142,20 +134,17 @@ const BaseEquipmentMenu = ({
                         <ItemViewInSpreadsheet
                             equipmentType={equipments.substations}
                             equipmentId={equipment.substationId}
-                            itemText={
-                                useName
-                                    ? getSubstationNameOrId(equipment)
-                                    : equipment.substationId
-                            }
+                            itemText={getNameOrId({
```
Comment (sBouzols) [inline src/components/menus/base-equipment-menu.js:137]:
```
itemText= {() => {
const substation = network.getSubstation(equipment?.substationId);
return getNameOrId(substation);
}}
```
even better for optimization create a little `useCallback`

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 16]
```diff
@@ -174,24 +178,18 @@ export function SingleLineDiagramPane({
...
     displayedSldHeightsRef.current = displayedSldHeights;
-
     const createView = useDisplayView(network, studyUuid, currentNode);
-
     const dispatch = useDispatch();
-
     const sldState = useSelector((state) => state.sldState);
-
     const viewsRef = useRef();
+    const { getNameOrId } = useNameOrId();
```
Comment (sBouzols) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:192]: seems not used, remove then ?

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 17]
```diff
@@ -121,20 +125,20 @@ const useDisplayView = (network, studyUuid, currentNode) => {
...
                 const vl = network.getVoltageLevel(vlId);
                 if (!vl) return;
-                let name = useName ? vl.name : vlId;
+                let label = getNameOrId({ name: vl?.name, id: vlId });
 
                 const substation = network.getSubstation(vlId);
                 const countryName = substation?.countryName;
                 if (countryName) {
-                    name += ' \u002D ' + countryName;
+                    label += ' \u002D ' + countryName;
```
Comment (sBouzols) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:133]: take care to use `' - '` when merging with main

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js [hunk 18]
```diff
@@ -100,19 +102,21 @@ const useDisplayView = (network, studyUuid, currentNode) => {
...
-                let name =
-                    useName && substation.name ? substation.name : substationId;
+                let label = getNameOrId({
+                    name: substation?.name,
+                    id: substationId,
+                });
                 const countryName = substation?.countryName;
                 if (countryName) {
-                    name += ' \u002D ' + countryName;
+                    label += ' \u002D ' + countryName;
```
Comment (sBouzols) [inline src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:111]: take care to use ' - ' when merging with main

File: src/components/util/equipmentInfosHandler.js (hunk 7 above)
Comment (sBouzols) [reply src/components/util/equipmentInfosHandler.js]: seems ok in your last revision

File: src/components/util/search-matching-equipments.js (hunk 9 above)
Comment (anistouri) [reply src/components/util/search-matching-equipments.js:45]: I changed in make Items so we must pass the label directly

File: src/utils/rest-api.js (hunk 11 above)
Comment (anistouri) [reply src/utils/rest-api.js:580]: it is a function it will be executed when the `urlSearchParams.append('fieldSelector', getUseNameParameterKey)` is called

File: src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js (hunk 16 above)
Comment (anistouri) [reply src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:192]: it's used in the line 367 and 399

Comment (sBouzols) [reply src/components/diagrams/singleLineDiagram/single-line-diagram-pane.js:191]: my bad... the previous function had the same name

File: src/components/util/search-matching-equipments.js (hunk 9 above)
Comment (sBouzols) [reply src/components/util/search-matching-equipments.js:45]: My bad, but it still doesn't work 
calling getNameOrId() without any parameter will always return null now
Here you want to pass the function `getNameOrId` to makeItems aka getEquipmentsInfosForSearchBar then 
`setEquipmentsFound(makeItems(infos, getNameOrId));`
and
```
var getEquipmentsInfosForSearchBar = function getEquipmentsInfosForSearchBar(equipmentsInfos, equipmentLabelling) {
  return equipmentsInfos.flatMap(function (e) {
    var label = equipmentLabelling(e);
    return e.type === 'SUBSTATION' ? [{
    ...
        type: e.type,
        voltageLevelLabel: equipmentLabelling(vli),
        voltageLevelId: vli.id
```

**Review by sBouzols — APPROVED** [review]
Code Review OK
Tests OK
console warning check OK
